devp2p/kademlia.py

```diff
@@ -421,22 +425,22 @@ def _expected_pongs():
         if pingid and (pingid not in self._expected_pongs):
             assert pingid not in self._expected_pongs
             log.debug('surprising pong', remoteid=node,
-                      expected=_expected_pongs(), pingid=pingid.encode('hex')[:8])
+                      expected=_expected_pongs(), pingid=encode_hex(pingid)[:8])
             if pingid in self._deleted_pingids:
                 log.debug('surprising pong was deleted')
             else:
                 for key in self._expected_pongs:
                     if key.endswith(node.pubkey):
                         log.debug('waiting for ping from node, but echo mismatch', node=node,
-                                  expected_echo=key[:len(node.pubkey)][:8].encode('hex'),
-                                  received_echo=pingid[:len(node.pubkey)][:8].encode('hex'))
+                                  expected_echo=encode_hex(key[:len(node.pubkey)][:8]),
+                                  received_echo=encode_hex(pingid[:len(node.pubkey)][:8]))
             return
 
         # check for timed out pings and eventually evict them
-        for _pingid, (timeout, _node, replacement) in self._expected_pongs.items():
+        for _pingid, (timeout, _node, replacement) in list(self._expected_pongs.items()):
```

Comment:
Same comment here about the mutation aspect.  I'd replace this with the following code.

```python
# Gather the id's and nodes that need to be removed.
pind_ids_to_delete = {
    _pingid: _node
    for _pingid, (timeout, _node, replacement)
    in self._expected_pongs.items()
    if time.time() > timeout
}
# replace the tracking dictionary with an updated one with the timeoud out items removed.
self._expected_pongs = {
    _pingid: (timeout, _node, replacement)
    for _pingid, (timeout, _node, replacement)
    in self._expected_pongs.items()
    if _pingid not in ping_ids_to_delete
}
# Log and execute the removals
for _pingid, _node in ping_ids_to_delete.items():
    self._deleted_pingids.add(_pingid)
    self.routing.remove_node(_node)
```